### DISCUSSION TITLE
Mineflayer не заходит на сервер, на тестовый локальный сервер он заходит

### DISCUSSION THREAD

**Discussion by Pos1t1veGuy** [PrismarineJS/mineflayer#2187] [Q&A]
Я создаю бота шахтера на основе примера из mineflayer. Я его тестил на локальном сервере, и на сервере aternos. Он туда заходил, и работал, но когда я стал его пробовать на других серверах, он ломается.
Он спокойно включается, думает, ждёт, и выключается. Он ни на что не ругается, просто выключается. 
Я знаю про анти-бот защиту на серверах, но я подбирал сервера без защиты. Я его научил регестрироваться, и с этим проблем нет.
Я перебирал пачку разных серверов, но на всех он ведёт себя одинаково. Была у него ошибка «ETIMEDOUT» но после шаманства в коде она пропала. Вот код: 
const mineflayer = require('mineflayer')
const AutoAuth = require('MineflayerAutoAuth')
const { pathfinder, Movements, goals: { GoalNear } } = require('mineflayer-pathfinder')
var navigatePlugin = require('mineflayer-navigate')(mineflayer);
var scaffoldPlugin = require('mineflayer-scaffold')(mineflayer);
 
const bot = mineflayer.createBot({
  host: 'mc.dexland.ru',
  port: 25565,
  username: 'Basya228',
  plugins: [AutoAuth],
  AutoAuth: '1230'
})
 
let target = null
 
navigatePlugin(bot);
scaffoldPlugin(bot);
bot.on('chat', function(username, message) {
  if (username === bot.username) return;
  var target1 = bot.players[username].entity;
  if (message === 'ко мне') {
    bot.scaffold.to(target1.position, function(err) {
      if (err) {
        bot.chat("Ой! Ошибка: " + err.code);
      } else {
        bot.chat("Бегу!");
      }
    });
  }
})
 
let wt
 
bot.on('chat', async (username, message) => {
  console.log(bot.entity.position.toString())
  if (username === bot.username) return
    target = bot.players[username].entity
    let entity
  bot.setControlState('forward', true)
  switch (message) {
    case 'загрузи чанки':
      await bot.waitForChunksToLoad()
      bot.chat('загрузил!')
      break
    case 'предметы':
      sayItems()
      break
    case 'копай под себя':
      dig()
      break
    case 'копай':
      dig4()
      setTimeout(dig1, 6000)
      setTimeout(path1, 12000)
      break
    case 'копай дальше':
      dig2()
      setTimeout(dig3, 6000)
      setTimeout(dig1, 12000)
      setTimeout(path1, 18000)
    case 'копай вперед':
      dig2()
      setTimeout(dig3, 6000)
      break
    case 'attack':
      entity = bot.nearestEntity()
      if (entity) {
        bot.attack(entity, true)
      } else {
        bot.chat('no nearby entities')
      }
    case 'поставь':
      build()
      break
    case 'землю':
      equipDirt()
      break
    case 'камень':
      equipStone()
      break
    case 'кирку':
      equipPickaxe()
      break
    case 'лопату':
      equipShovel()
      break
    case 'воду':
      equipWater()
      break
    case 'за мной':
      bot.setControlState('forward', true)
      break
    case 'назад':
      bot.setControlState('back', true)
      break
    case 'влево':
      bot.setControlState('left', true)
      break
    case 'вправо':
      bot.setControlState('right', true)
      break
    case 'беги':
      bot.setControlState('sprint', true)
      break
    case 'стой':
      bot.setControlState('forward', false)
      break
    case 'стоп':
      bot.clearControlStates()
      break
    case 'прыжок':
      bot.setControlState('jump', true)
      bot.setControlState('jump', false)
      break
    case 'прыгай':
      bot.setControlState('jump', true)
      break
    case 'не прыгай':
      bot.setControlState('jump', false)
      break
    case 'mount':
      entity = bot.nearestEntity((entity) => { return entity.type === 'object' })
      if (entity) {
        bot.mount(entity)
      } else {
        bot.chat('no nearby objects')
      }
      break
    case 'dismount':
      bot.dismount()
      break
    case 'move vehicle forward':
      bot.moveVehicle(0.0, 1.0)
      break
    case 'move vehicle backward':
      bot.moveVehicle(0.0, -1.0)
      break
    case 'move vehicle left':
      bot.moveVehicle(1.0, 0.0)
      break
    case 'move vehicle right':
      bot.moveVehicle(-1.0, 0.0)
      break
    case 'прыгни':
      bot.entity.position.y += 10
      break
    case 'pos':
      bot.chat(bot.entity.position.toString())
      break
    case 'yp':
      bot.chat(`Yaw ${bot.entity.yaw}, pitch: ${bot.entity.pitch}`)
      break
    case 'ты молодец':
      bot.chat(`спасибо, хозяин!`)
      break
    case 'вперед':
      path1()
      break
  }
})
 
bot.on('mount', () => {
  bot.chat(`mounted ${bot.vehicle.objectType}`)
})
 
bot.on('dismount', (vehicle) => {
  bot.chat(`dismounted ${vehicle.objectType}`)
})
 
function path1 () {
  var target2 = bot.blockAt(bot.entity.position.offset(0, -1, -1))
  bot.scaffold.to(target2.position, function(err) {
    if (err) {
      bot.chat("Ой! Ошибка: " + err.code)
      }
    })
  }
 
function sayItems (items = bot.inventory.items()) {
  const output = items.map(itemToString).join(', ')
  if (output) {
    bot.chat(output)
  } else {
    bot.chat('чет сломалось')
  }
}
 
function dig () { 
  let target
  if (bot.targetDigBlock) {
    bot.chat(`не могу вскопать ${bot.targetDigBlock.name}`)
    }   
  else {
    target = bot.blockAt(bot.entity.position.offset(0, -1, 0))
    if (target && bot.canDigBlock(target)) {
      bot.chat(`начинаю копать ${target.name}`)
      bot.dig(target, bot.chat(`я вскопал ${target.name}`))
      }
    }
}    
 
function dig1 () {
  if (bot.targetDigBlock) {
    bot.chat(`не могу вскопать ${bot.targetDigBlock.name}`)
    }   
  else {
    target1 = bot.blockAt(bot.entity.position.offset(0, -1, -1))
    if (target1 && bot.canDigBlock(target1)) {
      bot.chat(`начинаю копать ${target1.name}`)
      bot.dig(target1, bot.chat(`я вскопал ${target1.name}`))
      count = 1
    }
    else { 
      bot.chat('чет сломалось')
      }
    }
  }
  
function dig4 () {
  if (bot.targetDigBlock) {
    bot.chat(`не могу вскопать ${bot.targetDigBlock.name}`)
    }   
  else {
    target1 = bot.blockAt(bot.entity.position.offset(0, 0, -1))
    if (target1 && bot.canDigBlock(target1)) {
      bot.chat(`начинаю копать ${target1.name}`)
      bot.dig(target1, bot.chat(`я вскопал ${target1.name}`))
      count = 1
    }
    else { 
      bot.chat('чет сломалось')
      }
    }
  } 
  
function dig2 () {
  let target
  if (bot.targetDigBlock) {
    bot.chat(`не могу вскопать ${bot.targetDigBlock.name}`)
    }   
  else {
    target = bot.blockAt(bot.entity.position.offset(0, 0, -1))
    if (target && bot.canDigBlock(target)) {
      bot.chat(`начинаю копать ${target.name}`)
      bot.dig(target, bot.chat(`я вскопал ${target.name}`))
      }
    }
}    
 
function dig3 () {
  if (bot.targetDigBlock) {
    bot.chat(`не могу вскопать ${bot.targetDigBlock.name}`)
    }   
  else {
    target1 = bot.blockAt(bot.entity.position.offset(0, 1, -1))
    if (target1 && bot.canDigBlock(target1)) {
      bot.chat(`начинаю копать ${target1.name}`)
      bot.dig(target1, bot.chat(`я вскопал ${target1.name}`))
      count = 1
    }
    else { 
      bot.chat('чет сломалось')
      }
    }
  }
 
function build () {
  const referenceBlock = bot.blockAt(bot.entity.position.offset(0, -1, 0))
  const jumpY = Math.floor(bot.entity.position.y) + 1.0
  bot.setControlState('jump', true)
  bot.on('move', placeIfHighEnough)
 
  let tryCount = 0
 
  function placeIfHighEnough () {
    if (bot.entity.position.y > jumpY) {
      bot.placeBlock(referenceBlock, vec3(0, 1, 0), (err) => {
        if (err) {
          tryCount++
          if (tryCount > 10) {
            bot.chat(err.message)
            bot.setControlState('jump', false)
            bot.removeListener('move', placeIfHighEnough)
            return
          }
          return
        }
        bot.setControlState('jump', false)
        bot.removeListener('move', placeIfHighEnough)
        bot.chat('поставил блок')
      })
    }
  }
}
 
function equipDirt () {
  const mcData = require('minecraft-data')(bot.version)
  let itemsByName
  if (bot.supportFeature('itemsAreNotBlocks')) {
    itemsByName = 'itemsByName'
  } else if (bot.supportFeature('itemsAreAlsoBlocks')) {
    itemsByName = 'blocksByName'
  }
  bot.equip(mcData[itemsByName].dirt.id, 'hand', (err) => {
    if (err) {
      bot.chat(`у меня нету земли: ${err.message}`)
    } else {
      bot.chat('взял землю')
    }
  })
}
 
function equipStone () {
  const mcData = require('minecraft-data')(bot.version)
  let itemsByName
  if (bot.supportFeature('itemsAreNotBlocks')) {
    itemsByName = 'itemsByName'
  } else if (bot.supportFeature('itemsAreAlsoBlocks')) {
    itemsByName = 'blocksByName'
  }
  bot.equip(mcData[itemsByName].stone.id, 'hand', (err) => {
    if (err) {
      bot.chat(`у меня нету камня: ${err.message}`)
    } else {
      bot.chat('взял камень')
    }
  })
}
 
function equipPickaxe () {
  const mcData = require('minecraft-data')(bot.version)
  let itemsByName
  itemsByName = 'itemsByName'
  bot.equip(mcData[itemsByName].diamond_pickaxe.id, 'hand', (err) => {
    if (err) {
      bot.chat(`у меня нету кирки: ${err.message}`)
    } else {
      bot.chat('взял кирку')
    }
  })
}
 
function equipShovel () {
  const mcData = require('minecraft-data')(bot.version)
  let itemsByName
  itemsByName = 'itemsByName'
  bot.equip(mcData[itemsByName].diamond_shovel.id, 'hand', (err) => {
    if (err) {
      bot.chat(`у меня нету лопаты: ${err.message}`)
    } else {
      bot.chat('взял лопату')
    }
  })
}
 
function itemToString (item) {
  if (item) {
    return `${item.name} ${item.count} штук`
  } else {
    return '(nothing)'
  }
}
 
bot.on('kicked', (reason, loggedIn) => console.log(reason, loggedIn))
bot.on('error', err => console.log(err))

**Comment by Grammf** [comment]
Я, конечно, не особо разбираюсь в нод жс, но кажется у тебя инет шалит и не успевает выполнять код

**Reply by Grammf** [reply]
точнее тупо не дожидается ответа от серва

**Reply by Pos1t1veGuy** [reply]
Это конечно тоже может быть, но это происходило постоянно даже если перезагрузить роутер. И я в самой игре спокойно могу зайти на тот самый сервер

**Comment by zfxstd** [comment]
На самом деле сервера кубиков это целое искусство. У каждого второго сервера свои моды, свои плагины, свои оверайды, нередко свой код который оверайдит ванильный протокол итд, из за чего майнфлаер зачастую собственно говоря и ломается, не говоря о разных защитах от ботов которые даже не всегда кикают бота. Поэтому остается только дебажить и искать конкретные пакеты, которые не сломаны сервером и работают согласно протоколу и надеяться что это исправит проблему
